examples/tqc/utils.py

```diff
@@ -0,0 +1,410 @@
+
+
+import torch
+import numpy as np
+from tensordict.nn import InteractionType, TensorDictModule
+from tensordict.nn.distributions import NormalParamExtractor
+from torch import nn, optim
+from torchrl.collectors import SyncDataCollector
+from torchrl.data import TensorDictPrioritizedReplayBuffer, TensorDictReplayBuffer
+from torchrl.data.replay_buffers.storages import LazyMemmapStorage
+from torchrl.envs import Compose, DoubleToFloat, EnvCreator, ParallelEnv, TransformedEnv
+from torchrl.envs.libs.gym import GymEnv, set_gym_backend
+from torchrl.envs.transforms import InitTracker, RewardSum, StepCounter
+from torchrl.envs.utils import ExplorationType, set_exploration_type
+from torchrl.modules import MLP, ProbabilisticActor, ValueOperator
+from torchrl.modules.distributions import TanhNormal
+from torchrl.objectives import SoftUpdate
+from torchrl.data import CompositeSpec
+from torchrl.objectives.common import LossModule
+from tensordict.tensordict import TensorDict, TensorDictBase
+from typing import Tuple
+
+
+# ====================================================================
+# Environment utils
+# -----------------
+
+
+def env_maker(task, device="cpu"):
+    with set_gym_backend("gym"):
+        return GymEnv(
+            task,
+            device=device,
+        )
+
+
+def apply_env_transforms(env, max_episode_steps=1000):
+    transformed_env = TransformedEnv(
+        env,
+        Compose(
+            InitTracker(),
+            StepCounter(max_episode_steps),
+            DoubleToFloat(),
+            RewardSum(),
+        ),
+    )
+    return transformed_env
+
+
+def make_environment(cfg):
+    """Make environments for training and evaluation."""
+    parallel_env = ParallelEnv(
+        cfg.collector.env_per_collector,
+        EnvCreator(lambda: env_maker(task=cfg.env.name)),
+    )
+    parallel_env.set_seed(cfg.env.seed)
+
+    train_env = apply_env_transforms(parallel_env, cfg.env.max_episode_steps)
+
+    eval_env = TransformedEnv(
+        ParallelEnv(
+            cfg.collector.env_per_collector,
+            EnvCreator(lambda: env_maker(task=cfg.env.name)),
+        ),
+        train_env.transform.clone(),
+    )
+    return train_env, eval_env
+
+
+# ====================================================================
+# Collector and replay buffer
+# ---------------------------
+
+
+def make_collector(cfg, train_env, actor_model_explore):
+    """Make collector."""
+    collector = SyncDataCollector(
+        train_env,
+        actor_model_explore,
+        init_random_frames=cfg.collector.init_random_frames,
+        frames_per_batch=cfg.collector.frames_per_batch,
+        total_frames=cfg.collector.total_frames,
+        device=cfg.collector.collector_device,
+    )
+    collector.set_seed(cfg.env.seed)
+    return collector
+
+
+def make_replay_buffer(
+        batch_size,
+        prb=False,
+        buffer_size=1_000_000,
+        buffer_scratch_dir="/tmp/",
```

Comment:
Better use a unique temporary folder. Using tmp means that 2 independent simultaneous processes running the training loop will write data at the same place and interfere

Reply:
More like the changes in https://github.com/pytorch/rl/pull/1635
If you have the scratch_dir defined as `"/tmp/"`, two runs executed on the same machine will write the buffers in the same file(s)! This is amazingly bad.
If you need to checkpoint your replay buffer, you should give it a file path and make sure it's unique (ie no other exp can have the same path). Otherwise don't provide a path at all, things will be written in a temporary handler that will be removed after exiting the run.

Reply:
Thank you for explaining, that makes sense. I copied the implementation of the make_replay_buffer method from the TD3 code and followed the changes in #1635 .